common/interpreter/wawaka_wasm/WawakaInterpreter.cpp

```diff
@@ -192,9 +198,19 @@ int32 WawakaInterpreter::evaluate_function(
     pe::ThrowIfNull(wasm_func, "Unable to locate the dispatch function");
 
     // might need to add a null terminator
-    uint32 argv[2];
-    argv[0] = (int32)wasm_runtime_module_malloc(wasm_module_inst, args.length() + 1);
-    argv[1] = (int32)wasm_runtime_module_malloc(wasm_module_inst, env.length() + 1);
+    uint32 argv[2], buf_offset0, buf_offset1;
+
+    argv[0] = buf_offset0 = (int32)wasm_runtime_module_malloc(wasm_module_inst, args.length() + 1);
+    if (argv[0] == 0) {
+        SAFE_LOG(PDO_LOG_ERROR, "module malloc failed for some reason");
+        return 0;
+    }
+
+    argv[1] = buf_offset1 = (int32)wasm_runtime_module_malloc(wasm_module_inst, env.length() + 1);
+    if (argv[1] == 0) {
+        SAFE_LOG(PDO_LOG_ERROR, "module malloc failed for some reason");
+        return 0;
+    }
 
```

Comment:
we definitely need to free the allocated buffers. would prefer to put this in a try/catch/finally block. for example, this code doesn't free buffer 0 if buffer 1 fails. and we generally throw exceptions on allocation errors. 

final question... do we need to free the result once we copy it out? presumably... since we destroy the interpreter at the end of the execution, everything is free'd anyway.

Reply:
Sorry for not freeing buffer 0 if buffer 1 fails, you should add it.

The buffer is allocated from the heap space, which is initialized int module instantiating and will be destroyed when module instance is de-instantiated. So no memory leak will happen if you don't free them, the issue is that, if you keep calling module_malloc, it may fail when the heap is used up. So it you copy the buffer out, you don't need to free it.

And if module_malloc fails, the runtime throws exception: out of memory.